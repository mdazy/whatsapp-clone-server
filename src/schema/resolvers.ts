import { DateTimeResolver, URLResolver } from "graphql-scalars";
import { chats, users, messages, chatRooms } from "../db";

const resolvers = {
  Date: DateTimeResolver,
  URL: URLResolver,
  Chat: {
    user(chat: any) {
      return users.filter(u => u.id === chat.userId)[0];
    },
    lastMessage(chat: any) {
      return messages.filter(m => m.id === chat.lastMessageId)[0];
    }
  },
  ChatRoom: {
    messages(chat: any) {
      const chatRoom = chatRooms.filter(c => c.id === chat.id)[0];
      return chatRoom.messageIds.map(
        id => messages.filter(mm => mm.id === id)[0]
      );
    }
  },
  Query: {
    chats() {
      return chats;
    },
    chat(_: any, params: any) {
      const id = Number(params.id);
      let chatRoom = chatRooms.filter(c => c.id === id);
      if (chatRoom.length === 0) {
        // hmmm which do we want?
        // throw new Error(`No chat with id ${id}`);
        return undefined;
      }
      return { id };
    }
  }
};

export default resolvers;
