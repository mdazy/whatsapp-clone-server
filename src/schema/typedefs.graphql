scalar Date
scalar URL

type Message {
  id: ID!
  fromUser: User!
  content: String!
  date: Date!
}

type User {
  id: ID!
  name: String!
  image: URL!
}

type Chat {
  id: ID!
  user: User!
  lastMessage: Message!
  messages: [Message!]!
}

type Query {
  chats: [Chat!]!
  chat(id: ID): Chat
}

type Mutation {
  sendMessage(chatId: ID!, userId: ID!, content: String!): Message
}
