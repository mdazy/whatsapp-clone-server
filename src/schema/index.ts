import { importSchema } from "graphql-import";
import { makeExecutableSchema } from "graphql-tools";
import resolvers from "./resolvers";

const typeDefs = importSchema("src/schema/typedefs.graphql");

export default makeExecutableSchema({ resolvers, typeDefs });
