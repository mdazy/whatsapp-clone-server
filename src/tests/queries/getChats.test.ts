import { createTestClient } from "apollo-server-testing";
import { ApolloServer } from "apollo-server-express";
import schema from "../../schema";

describe("Query.chats", () => {
  it("should fetch all chats", async () => {
    const server = new ApolloServer({ schema });

    const { query } = createTestClient(server);

    const res = await query({
      query: `
        query {
            chats {
              id
              user {
                name
                image
              }
              lastMessage {
                content
                date
              }
            }
          }`
    });

    expect(res.data).toBeDefined();
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });
});
