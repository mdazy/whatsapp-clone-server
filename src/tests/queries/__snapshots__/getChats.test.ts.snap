// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query.chats should fetch all chats 1`] = `
Object {
  "chats": Array [
    Object {
      "id": "1",
      "lastMessage": Object {
        "content": "Hi there",
        "date": "2019-09-25T07:11:28.693Z",
      },
      "user": Object {
        "image": "https://randomuser.me/api/portraits/thumb/women/58.jpg",
        "name": "Alice",
      },
    },
    Object {
      "id": "2",
      "lastMessage": Object {
        "content": "Be right back",
        "date": "2019-09-24T22:11:28.693Z",
      },
      "user": Object {
        "image": "https://randomuser.me/api/portraits/thumb/men/58.jpg",
        "name": "Bob",
      },
    },
    Object {
      "id": "3",
      "lastMessage": Object {
        "content": "Let's get this done",
        "date": "2019-09-25T05:11:28.693Z",
      },
      "user": Object {
        "image": "https://randomuser.me/api/portraits/thumb/men/42.jpg",
        "name": "Carl",
      },
    },
  ],
}
`;
